functions-and-operators/string-functions.md

```diff
@@ -1268,22 +1268,343 @@ Escape the argument for use in an SQL statement.
++-------------------------+
+|                       1 |
++-------------------------+
+1 row in set (0.00 sec)
+```
+
+This example matches `^A` against `abc`, which doesn't match.
+
+```sql
+SELECT REGEXP_LIKE('abc','^A');
+```
+
+```
++-------------------------+
+| REGEXP_LIKE('abc','^A') |
++-------------------------+
+|                       0 |
++-------------------------+
+1 row in set (0.00 sec)
+```
+
+This example matches `^A` against `abc`, which now matches because of the `i` flag which enabled case insensitive matching.
+
+```sql
+SELECT REGEXP_LIKE('abc','^A','i');
+```
+
+```
++-----------------------------+
+| REGEXP_LIKE('abc','^A','i') |
++-----------------------------+
+|                           1 |
++-----------------------------+
+1 row in set (0.00 sec)
+```
+
 ### [`REGEXP_REPLACE()`](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-replace)
 
 Replace substrings that match the regular expression (Partly compatible with MySQL. For more details, see [Regular expression compatibility with MySQL](#regular-expression-compatibility-with-mysql)).
 
+The `REGEXP_REPLACE(str, regexp, replace, [start, [match, [flags]]])` function can be used to replace strings based on regular expressions.
+
+Examples:
+
+In this example two o's are replaced by "i".
+
+```sql
+SELECT REGEXP_REPLACE('TooDB', 'o{2}', 'i');
+```
+
+```
++--------------------------------------+
+| REGEXP_REPLACE('TooDB', 'o{2}', 'i') |
++--------------------------------------+
+| TiDB                                 |
++--------------------------------------+
+1 row in set (0.00 sec)
+```
+
+In the example below we start the match at the third character, causing the regular expression to not match and not do any replacement.
+
+```sql
+SELECT REGEXP_REPLACE('TooDB', 'o{2}', 'i',3);
+```
+
+```
++----------------------------------------+
+| REGEXP_REPLACE('TooDB', 'o{2}', 'i',3) |
++----------------------------------------+
+| TooDB                                  |
++----------------------------------------+
+1 row in set (0.00 sec)
+```
+
+In the example below the 5th argument is used to set if the first or the second match is used for the replacement.
+
+```sql
+SELECT REGEXP_REPLACE('TooDB', 'o', 'i',1,1);
+```
+
+```
++---------------------------------------+
+| REGEXP_REPLACE('TooDB', 'o', 'i',1,1) |
++---------------------------------------+
+| TioDB                                 |
++---------------------------------------+
+1 row in set (0.00 sec)
+```
+
+```sql
+SELECT REGEXP_REPLACE('TooDB', 'o', 'i',1,2);
+```
+
+```
++---------------------------------------+
+| REGEXP_REPLACE('TooDB', 'o', 'i',1,2) |
++---------------------------------------+
+| ToiDB                                 |
++---------------------------------------+
+1 row in set (0.00 sec)
+```
+
+In the example below we use the 6th argument to set the flags for case insensitive matching.
+
+```sql
+SELECT REGEXP_REPLACE('TooDB', 'O{2}','i',1,1);
```

Comment:
Btw, are these expected results? (same in both MySQL and TiDB)
```
tidb> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',1);
+---------------------------------------+
| REGEXP_REPLACE('ToooDB', 'oo', 'i',1) |
+---------------------------------------+
| TioDB                                 |
+---------------------------------------+
1 row in set (0.00 sec)

tidb> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',2);
+---------------------------------------+
| REGEXP_REPLACE('ToooDB', 'oo', 'i',2) |
+---------------------------------------+
| TioDB                                 |
+---------------------------------------+
1 row in set (0.00 sec)

tidb> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',3);
+---------------------------------------+
| REGEXP_REPLACE('ToooDB', 'oo', 'i',3) |
+---------------------------------------+
| ToiDB                                 |
+---------------------------------------+
1 row in set (0.00 sec)
```
and
```
mysql> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',0);
ERROR 1583 (42000): Incorrect parameters in the call to native function 'regexp_replace'
mysql> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i');
+-------------------------------------+
| REGEXP_REPLACE('ToooDB', 'oo', 'i') |
+-------------------------------------+
| TioDB                               |
+-------------------------------------+
1 row in set (0,00 sec)
```

Reply:
The first three queries in your comment look correct to me.  For the last two queries MySQL and TiDB don't have the same error message:


MySQL 8.3.0:
```
mysql> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',0);
ERROR 1583 (42000): Incorrect parameters in the call to native function 'regexp_replace'
```

TiDB 8.0.0:
```
tidb> SELECT REGEXP_REPLACE('ToooDB', 'oo', 'i',0);
ERROR 1139 (42000): Got error 'Index out of bounds in regular expression search' from regexp
```

Note that 1 is the default for the 4th argument and values less than that are invalid.

Reply:
Ah, sorry for missing the index starts at 1 :)
I just didn't see how 'XoooY' would match/replace the regex 'oo' for both 1st and 2nd match the first occurrence of 'o', and the 3rd match the second occurrence of 'o'.
But since it is the same in MySQL I don't mind why :)